pkg/operator/resource/resourceapply/core.go

```diff
@@ -131,14 +134,41 @@ func ApplyConfigMap(client coreclientv1.ConfigMapsGetter, recorder events.Record
 
 	modified := resourcemerge.BoolPtr(false)
 	resourcemerge.EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
-	dataSame := equality.Semantic.DeepEqual(existing.Data, required.Data)
+
+	var modifiedKeys []string
+	for k, v := range existing.Data {
+		if v2, ok := required.Data[k]; !ok || !equality.Semantic.DeepEqual(v, v2) {
+			modifiedKeys = append(modifiedKeys, "data."+k)
+		}
+	}
+	for k := range required.Data {
+		if _, ok := existing.Data[k]; !ok {
+			modifiedKeys = append(modifiedKeys, "data."+k)
+		}
+	}
+
+	dataSame := len(modifiedKeys) == 0
 	if dataSame && !*modified {
 		return existing, false, nil
 	}
 	existing.Data = required.Data
 
 	actual, err := client.ConfigMaps(required.Namespace).Update(existing)
-	reportUpdateEvent(recorder, required, err)
+
+	namespace := required.GetNamespace()
+	if len(namespace) > 0 {
+		namespace = "." + namespace
+	}
+	if err != nil {
+		recorder.Warningf("ConfigMapFailed", "Failed to update ConfigMap/%s%s: %v", required.GetName(), namespace, err)
+		return actual, true, err
+	}
+	if len(modifiedKeys) == 0 {
+		recorder.Eventf("ConfigMapUpdated", "Updated ConfigMap/%s%s because it changed: %v", required.GetName(), namespace)
+	} else {
+		sort.Sort(sort.StringSlice(modifiedKeys))
+		recorder.Eventf("ConfigMapUpdated", "Updated ConfigMap/%s%s because the following fields changed: %v", required.GetName(), namespace, strings.Join(modifiedKeys, ","))
```

Comment:
then you can use the `"caused by changes in %v", strings.Join(modifiedKeys****` to pass to it